wallet/controllers.go

```diff
@@ -0,0 +1,204 @@
+type LinkWalletRequest struct {
+	SignedTx         string     `json:"signedTx"`
+	AnonymousAddress *uuid.UUID `json:"anonymousAddress"`
+}
+
+// PostLinkWalletCompat links wallets using provided ids
+func PostLinkWalletCompat(service *Service) handlers.AppHandler {
+	return handlers.AppHandler(func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		paymentIDString := chi.URLParam(r, "paymentID")
+		paymentID, err := uuid.FromString(paymentIDString)
+		if err != nil {
+			return handlers.ValidationError("url parameter", map[string]string{
+				"paymentID": "must be a valid uuidv4",
+			})
+		}
+
+		var body LinkWalletRequest
+		err = requestutils.ReadJSON(r.Body, &body)
+		if err != nil {
+			return handlers.ValidationError("request body", map[string]string{
+				"body": "unable to ready body",
+			})
+		}
+		_, err = govalidator.ValidateStruct(body)
+		if err != nil {
+			return handlers.WrapValidationError(err)
+		}
+
+		wallet, err := service.ReadableDatastore().GetWallet(paymentID)
+		if err != nil {
+			return handlers.WrapError(err, "Error finding wallet", http.StatusNotFound)
+		}
+		err = service.LinkWallet(r.Context(), wallet, body.SignedTx, body.AnonymousAddress)
+		if err != nil {
+			return handlers.WrapError(err, "error linking wallet", http.StatusBadRequest)
+		}
+
+		return handlers.RenderContent(r.Context(), wallet, w, http.StatusOK)
+	})
+}
+
+// PostCreateWalletResponse includes a ClaimID which can later be used to check the status of the claim
+type PostCreateWalletResponse struct {
+	Wallet     *walletutils.Info `json:"wallet"`
+	PrivateKey *string           `json:"privateKey"`
+}
+
+// GetWalletResponse gets wallet info tied to a wallet id
+type GetWalletResponse struct {
+	Wallet *walletutils.Info `json:"wallet"`
+}
+
+// PostCreateWalletRequest has possible inputs for the new wallet
+type PostCreateWalletRequest struct {
+	Provider  string `json:"provider" valid:"in(brave|uphold)"`
+	PublicKey string `json:"publicKey" valid:"required"`
+	SignedTx  string `json:"signedTx" valid:"-"`
+}
+
+// PostCreateWallet creates a wallet
+func PostCreateWallet(service *Service) handlers.AppHandler {
+	return handlers.AppHandler(func(w http.ResponseWriter, r *http.Request) *handlers.AppError {
+		limit := int64(1024 * 1024 * 10) // 10MiB
+		body, err := ioutil.ReadAll(io.LimitReader(r.Body, limit))
+		if err != nil {
+			return handlers.WrapError(err, "Error reading body", http.StatusBadRequest)
+		}
+
+		var req PostCreateWalletRequest
+		err = json.Unmarshal(body, &req)
+		if err != nil {
+			return handlers.WrapError(err, "Error unmarshalling body", http.StatusBadRequest)
+		}
+		_, err = govalidator.ValidateStruct(req)
+		if err != nil {
+			return handlers.WrapValidationError(err)
+		}
+
+		publicKey, err := middleware.GetKeyID(r.Context())
+		if err != nil {
+			return handlers.WrapError(err, "unable to look up http signature info", http.StatusBadRequest)
+		}
+		if req.PublicKey != publicKey {
```

Comment:
If we have this public key already from HTTPSignedOnly, then why do we need it in the request too?  Can we just use the public key we lookup in the HTTPSignedOnly middleware?

Reply:
[67ceddd](https://github.com/brave-intl/bat-go/pull/314/commits/67ceddd009a50507eace248b3a81c789b287defd)